core/libp2p/multi/multiaddress.hpp

```diff
@@ -6,11 +6,136 @@
 #ifndef KAGOME_MULTIADDRESS_HPP
 #define KAGOME_MULTIADDRESS_HPP
 
+#include <memory>
+#include <string>
+#include <string_view>
+
+#include <boost/optional.hpp>
+#include "common/buffer.hpp"
+#include "common/result.hpp"
+
 namespace libp2p::multi {
   /**
    * Address format, used by Libp2p
    */
-  class Multiaddress {};
+  class Multiaddress {
+   private:
+    using ByteBuffer = kagome::common::Buffer;
+    using FactoryResult =
+        kagome::expected::Result<std::unique_ptr<Multiaddress>, std::string>;
+
+   public:
+    Multiaddress() = delete;
+
+    Multiaddress(const Multiaddress &address) = default;
+    Multiaddress &operator=(const Multiaddress &address) = default;
+
+    Multiaddress(Multiaddress &&address) = default;
+    Multiaddress &operator=(Multiaddress &&address) = default;
+
+    /**
+     * Construct a multiaddress instance from the string
+     * @param address - string to be in that multiaddress
+     */
+    static FactoryResult createMultiaddress(std::string_view address);
+
+    /**
+     * Construct a multiaddress instance from the bytes
+     * @param bytes to be in that multiaddress
+     */
+    static FactoryResult createMultiaddress(std::shared_ptr<ByteBuffer> bytes);
```

Comment:
don't use pointer where it is not necessary. you could use `const ByteBuffer& bytes`